src/Build/Evaluation/ItemDataCollectionValue.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Build.Evaluation
+{
+    /// <summary>
+    /// An efficient multi-value wrapper holding one or more items.
+    /// </summary>
+    internal struct ItemDataCollectionValue<I>
+    {
+        /// <summary>
+        /// A non-allocating enumerator for the multi-value.
+        /// </summary>
+        public struct Enumerator : IEnumerator<I>
+        {
+            private object _value;
+            private int _index;
+
+            public Enumerator(object value)
+            {
+                _value = value;
+                _index = -1;
+            }
+
+            public I Current => (_value is IList<I> list) ? list[_index] : (I)_value;
+            object System.Collections.IEnumerator.Current => Current;
+
+            public void Dispose()
+            { }
+
+            public bool MoveNext()
+            {
+                // If value is not a list, it is a single item.
+                int count = (_value is IList<I> list) ? list.Count : 1;
+                if (_index + 1 < count)
+                {
+                    _index++;
+                    return true;
+                }
+                return false;
+            }
+
+            public void Reset()
+            {
+                _index = -1;
+            }
+        }
+
+        /// <summary>
+        /// Holds one value or a list of values.
+        /// </summary>
+        private object _value;
+
+        public bool IsEmpty => _value == null || (_value is List<I> list && list.Count == 0);
+
+        public ItemDataCollectionValue(I item)
+        {
+            _value = item;
+        }
+
+        public void Add(I item)
+        {
+            if (_value is not List<I> list)
+            {
+                list = new List<I>();
+                _value = list;
+            }
+            list.Add(item);
```

Comment:
Should we have some unit tests for this class? My first instinct is that this looks like it would drop a single item when adding the next one.

Reply:
Good catch! We definitely don't have this covered with tests, will fix.

Reply:
I have fixed the code, added a unit test, and made `ProjectItem_Tests` run with as well as without the directory-based optimization.